hal/src/sercom/v2/uart/mod.rs

```diff
@@ -0,0 +1,1528 @@
+// Uart
+//=============================================================================
+
+/// Abstraction over a UART peripheral, allowing to perform UART transactions.
+/// The second type parameter, `D`, denotes what the struct's [`Capability`] is.
+///
+/// * [`Rx`] or [`RxDuplex`]: Can perform receive transactions
+/// * [`Tx`] or [`TxDuplex`]: Can perform transmit transactions
+/// * [`Duplex`]: Can perform receive and transmit transactions. Additionally,
+///   you can call [`split`](Uart::split) to return a `(Uart<C, RxDuplex>,
+///   Uart<C, TxDuplex>)` tuple.
+pub struct Uart<C, D>
+where
+    C: ValidConfig,
+    D: Capability,
+{
+    config: C,
+    capability: PhantomData<D>,
+}
+
+#[cfg(any(feature = "samd11", feature = "samd21"))]
+type USART = pac::sercom0::USART;
+
+#[cfg(feature = "min-samd51g")]
+type USART = pac::sercom0::USART_INT;
+
+impl<C, D> Uart<C, D>
+where
+    C: ValidConfig,
+    D: Capability,
+{
+    /// Helper method to obtain a reference to given `USART` from the PAC
+    /// `SERCOM` struct
+    #[inline]
+    pub(crate) fn usart(&self) -> &USART {
+        self.config.as_ref().registers.usart()
+    }
+
+    /// Helper method to remove the interrupt flags not pertinent to `Self`'s
+    /// `Capability`
+    fn capability_flags(flags: Flags) -> Flags {
+        flags & unsafe { Flags::from_bits_unchecked(D::FLAG_MASK) }
+    }
+
+    /// Helper method to remove the status flags not pertinent to `Self`'s
+    /// `Capability`
+    fn capability_status(status: Status) -> Status {
+        status & unsafe { Status::from_bits_unchecked(D::STATUS_MASK) }
+    }
+
+    /// Read the interrupt flags
+    pub fn read_flags(&self) -> Flags {
+        let bits = self.usart().intflag.read().bits();
+        Flags::from_bits_truncate(bits)
+    }
+
+    /// Clear interrupt status flags
+    ///
+    /// Setting the `ERROR`, `RXBRK`, `CTSIC`, `RXS`, or `TXC` flag will clear
+    /// the interrupts. This function has no effect on the `DRE` or
+    /// `RXC` flags.
+    ///
+    /// Note that only the flags pertinent to `Self`'s [`Capability`]
+    /// will be cleared. The other flags will be **SILENTLY IGNORED**.
+    ///
+    /// * Available flags for [`Receive`] capability: `RXC`, `RXS`, `RXBRK` and
+    ///   `ERROR`
+    /// * Available flags for [`Transmit`] capability: `DRE`, `TXC` and `CTSIC`
+    /// * Since [`Duplex`] [`Uart`]s are [`Receive`] + [`Transmit`] they have
+    ///   all flags available.
+    pub const TX_FLAG_MASK: u8 = DRE | TXC | CTSIC;
+    ///
+    /// **Warning:** The implementation of of [`Write::flush`] waits on and
+    /// clears the `TXC` flag. Manually clearing this flag could cause it to
+    /// hang indefinitely.
+    pub fn clear_flags(&mut self, flags: Flags) {
+        // Remove flags not pertinent to Self's Capability
+        self.usart()
+            .intflag
+            .modify(|_, w| unsafe { w.bits(Self::capability_flags(flags).bits()) });
+    }
```

Comment:
You might consider moving these functions to `Registers` as well. I'm not sure if you agree, but I see value in having **all** of the register interaction occur in one and *only* one type.

That being said, it's mostly a preference.

In this particular case, I think you could take the `flags` argument and intersect it with the `.capability_flags()` value before passing the corresponding bits to `self.config.as_mut().registers.clear_flags()`.